proposals/0022-validating-resource-container-elements.md

```diff
@@ -0,0 +1,123 @@
+The compiler will need to analyze resource element types to determine whether the element type
+is valid and acceptable. This includes user-defined types. Ideally, a user should be able to
+determine how any user-defined structure is invalid as a resource element type. 
+
+## Proposed solution
+
+The proposed solution is to create a dedicated function to determine whether a given type
+is a valid resource element type. The function will  recursively examine the given input
+type, and determine whether all leaf element types are the set of acceptable primitives
+(i.e., int, uint, half, float, or double.) If, for example, the resource element type is
+an array of other resources, an error should be emitted.
+
+## Detailed design
+
+Initially, the function will start with the input resource element type. Due to the nature
+of the problem, the function can be recursively reapplied to any vector or matrix-like structures
+within the input type. Formally, the resource element type is valid if the function returns true
+on the given input type, or if the function returns true on all fields within the given input type.
+
+Firstly, the base case will be checked. The given input type will be compared to the set of acceptable
+primitives that are valid for a resource element. (i.e., float, int, etc).
+If there are no matches, then the resource element type may be an array. In that case, the function
+checks the array element type, and if it contains an element type in the set of acceptable primitives,
+then the array is valid.
+If the resource element was not an array, then the resource element may be a struct. In that case,
+the compiler will iterate through all fields of the struct, which may contain primitive members,
+arrays, or other structs. The function will be recursively called on each field of the struct,
+verifying that the leaf types are valid primitive elements.
+If the function never returns false at any point, then the type is valid.
+
+
+### HLSL Additions
+* How is this feature represented in the grammar?
+This feature won't be represented in the language.
+* How does it interact with different shader stages?
+The feature is independent of shader stages, since resource type validation
+doesn't depend on the shader stage.
+* How does it interact other HLSL features (semantics, buffers, etc)?
+The implementation of this feature will add diagnostics on resources.
+* How does this interact with C++ features that aren't already in HLSL?
+No interaction.
+* Does this have implications for existing HLSL source code compatibility?
+Possibly, if we decide that the valid set of primitive element types is different
+than what was decided with DXC, or if the validation method significantly differs
+from DXC.
+
+### Interchange Format Additions
+
+* What DXIL changes does this change require?
+None
+* What Metadata changes does this require?
+None
+* How will SPIRV be supported?
+No need

```

Comment:
As these questions aren't really relevant, I wonder if this should really be using the infra template?